src/wristpy/processing/analytics.py

```diff
@@ -0,0 +1,305 @@
+            if min_onset is not None and max_wakeup is not None:
+                onset.append(min_onset)
+                wakeup.append(max_wakeup)
+
+        return SleepWindow(onset=onset, wakeup=wakeup)
+
+    def _find_long_blocks(
+        self, below_threshold: np.ndarray, block_length: int = 360
+    ) -> np.ndarray:
+        """Helper function to find long blocks where SPT window is true.
+
+        This function uses the convolution of a kernel of 1s, of length block_length,
+        to find the continuous long blocks where SPT window is true. Where the
+        convolution value is == long_block_length that implies an overlap between the
+        kernel and the threshold signal of length long_block.
+
+        Args:
+            below_threshold: the 5-minute rolling median of the anglez difference
+                that is true when below the cutoff threshold.
+            block_length: the length of the long block that defines sleep, default is
+                30 minutes. (360 chunks of 5s)
+
+        Returns:
+            A numpy array with 1s indicating the identified SPT windows.
+        """
+        kernel = np.ones(block_length, dtype=int)
+        convolved = np.convolve(below_threshold, kernel, mode="same")
+        long_blocks_idx = np.where(convolved == block_length)[0]
+        sleep_idx_array = np.zeros(len(below_threshold))
+        sleep_idx_array[long_blocks_idx] = 1
+
+        return sleep_idx_array
+
+    def _fill_short_blocks(
+        self, sleep_idx_array: np.ndarray, gap_block: int = 720
+    ) -> np.ndarray:
+        """Helper function to fill gaps in SPT window that are less than 60 minutes.
+
+        We first find all the zeros in the sleep_idx_array, and then use np.diff to find
+        the gaps between those zeros, if np.diff !=0, that implies there is a gap
+        between 0s. We split slp_idx_array and find, and fill, the blocks that
+        are < gap_block.
+
+        Args:
+            sleep_idx_array: the array of SPT windows.
+            gap_block: the length of the gap that defines sleep, default is 60 minutes.
+                The units are chunks of 5s.
+
+        Returns:
+            A numpy array with 1s indicating the identified SPT windows.
+        """
+        zeros_idx = np.where(sleep_idx_array == 0)[0]
+        n_blocks = np.split(zeros_idx, np.where(np.diff(zeros_idx) != 1)[0] + 1)
+        for block_idx, block in enumerate(n_blocks):
+            if block_idx != 0 and len(block) < (gap_block):
+                sleep_idx_array[block] = 1
+
+        return sleep_idx_array
+
+    def _compute_abs_diff_mean_anglez(
+        self, anglez_data: models.Measurement, window_size_seconds: int = 5
+    ) -> models.Measurement:
+        """Helper function to compute the absolute difference of averaged anglez data.
+
+        Args:
+            anglez_data: the raw anglez data.
+                ##Note if we have access to the epoch1 anglez data this function will be
```

Comment:
Just noticed I missed this ##Note. If we need to keep track of this, put it in an issue along with the requirements for getting epoch1 anglez data.

Reply:
Unfortunately I'm not the Eye of Sauron; I don't see everything :'(